rust-runtime/smithy-json/src/escape.rs

```diff
@@ -5,35 +5,38 @@
 
 use std::borrow::Cow;
 
-const ESCAPES: &[char] = &['"', '\\', '\u{08}', '\u{0C}', '\n', '\r', '\t'];
-
 /// Escapes a string for embedding in a JSON string value.
 pub fn escape_string(value: &str) -> Cow<str> {
-    if !value.contains(ESCAPES) {
-        return Cow::Borrowed(value);
+    let bytes = value.as_bytes();
+    for (index, byte) in bytes.iter().enumerate() {
+        match byte {
+            0..=0x1F | b'"' | b'\\' => {
+                return Cow::Owned(escape_string_inner(&bytes[0..index], &bytes[index..]))
+            }
+            _ => {}
+        }
     }
+    Cow::Borrowed(value)
+}
 
-    let mut escaped = String::new();
-    let (mut last, end) = (0, value.len());
-    for (index, chr) in value
-        .char_indices()
-        .filter(|(_index, chr)| ESCAPES.contains(chr))
-    {
-        escaped.push_str(&value[last..index]);
-        escaped.push_str(match chr {
-            '"' => "\\\"",
-            '\\' => "\\\\",
-            '\u{08}' => "\\b",
-            '\u{0C}' => "\\f",
-            '\n' => "\\n",
-            '\r' => "\\r",
-            '\t' => "\\t",
-            _ => unreachable!(),
-        });
-        last = index + 1;
+fn escape_string_inner(start: &[u8], rest: &[u8]) -> String {
+    let mut escaped = start.to_vec();
+    for byte in rest {
+        match byte {
+            b'"' => escaped.extend("\\\"".bytes()),
+            b'\\' => escaped.extend("\\\\".bytes()),
+            0x08 => escaped.extend("\\b".bytes()),
+            0x0C => escaped.extend("\\f".bytes()),
+            b'\n' => escaped.extend("\\n".bytes()),
+            b'\r' => escaped.extend("\\r".bytes()),
+            b'\t' => escaped.extend("\\t".bytes()),
+            0..=0x1F => escaped.extend(format!("\\u{:04x}", byte).bytes()),
+            _ => escaped.push(*byte),
+        }
     }
-    escaped.push_str(&value[last..end]);
-    Cow::Owned(escaped)
+    // Our input was originally valid UTF-8, and we didn't do anything to invalidate it
```

Comment:
the safety argument is actually a little more subtle here:
- We only escaped bytes that were already single character code points
- We only replaced them with valid UTF-8